modules/core/include/opencv2/core/dualquaternion.inl.hpp

```diff
@@ -0,0 +1,392 @@
+}
+
+template <typename T>
+inline DualQuat<T>& DualQuat<T>::operator-=(const DualQuat<T> &q)
+{
+    *this = *this - q;
+    return *this;
+}
+
+template <typename T>
+inline DualQuat<T> operator-(const T a, const DualQuat<T> &q)
+{
+    return DualQuat<T>(a - q.w, -q.x, -q.y, -q.z, -q.w_, -q.x_, -q.y_, -q.z_);
+}
+
+template <typename T>
+inline DualQuat<T> operator*(const T a, const DualQuat<T> &q)
+{
+    return DualQuat<T>(q.w * a, q.x * a, q.y * a, q.z * a, q.w_ * a, q.x_ * a, q.y_ * a, q.z_ * a);
+}
+
+template <typename T>
+inline DualQuat<T> operator*(const DualQuat<T> &q, const T a)
+{
+    return DualQuat<T>(q.w * a, q.x * a, q.y * a, q.z * a, q.w_ * a, q.x_ * a, q.y_ * a, q.z_ * a);
+}
+
+template <typename T>
+inline DualQuat<T> DualQuat<T>::operator/(const T a) const
+{
+    return DualQuat<T>(w / a, x / a, y / a, z / a, w_ / a, x_ / a, y_ / a, z_ / a);
+}
+
+template <typename T>
+inline DualQuat<T> DualQuat<T>::operator/(const DualQuat<T> &q) const
+{
+    return *this * q.inv();
+}
+
+template <typename T>
+inline DualQuat<T>& DualQuat<T>::operator/=(const DualQuat<T> &q)
+{
+    *this = *this / q;
+    return *this;
+}
+
+template <typename T>
+std::ostream & operator<<(std::ostream &os, const DualQuat<T> &q)
+{
+    os << "DualQuat " << Vec<T, 8>{q.w, q.x, q.y, q.z, q.w_, q.x_, q.y_, q.z_};
+    return os;
+}
+
+template <typename T>
+DualQuat<T> DualQuat<T>::exp() const
+{
+    DualQuat<T> v(0, x, y, z, 0, x_, y_, z_);
+    DualQuat<T> normV = v.norm();
+    DualQuat<T> sin_normV(std::sin(normV.w), 0, 0, 0, normV.w_ * std::cos(normV.w), 0, 0, 0);
+    DualQuat<T> cos_normV(std::cos(normV.w), 0, 0, 0, -normV.w_ * std::sin(normV.w), 0, 0, 0);
+    DualQuat<T> exp_w(std::exp(w), 0, 0, 0, w_ * std::exp(w), 0, 0, 0);
+    if (normV.w < CV_DUAL_QUAT_EPS)
+    {
+        return exp_w * (cos_normV + v);
+    }
+    DualQuat<T> k = sin_normV / normV;
+    return exp_w * (cos_normV + v * k);
+}
+
+template <typename T>
+DualQuat<T> DualQuat<T>::log(QuatAssumeType assumeUnit) const
+{
+    DualQuat<T> v(0, x, y, z, 0, x_, y_, z_);
+    DualQuat<T> normV = v.norm();
+    if (assumeUnit)
+    {
+        if (normV.w < CV_DUAL_QUAT_EPS)
+        {
+            return v;
+        }
+        DualQuat<T> k = DualQuat<T>(std::acos(w), 0, 0, 0, -w_ * 1.0 / std::sqrt(1 - w * w), 0, 0, 0) / normV;
+        return v * k;
+    }
+    DualQuat<T> qNorm = norm();
+    if (qNorm.w < CV_DUAL_QUAT_EPS)
+    {
+        CV_Error(Error::StsBadArg, "Cannot apply this quaternion to log function: undefined");
+    }
+    DualQuat<T> log_qNorm(std::log(qNorm.w), 0, 0, 0, qNorm.w_ / qNorm.w, 0, 0, 0);
+    if (normV.w < CV_DUAL_QUAT_EPS)
+    {
+        return log_qNorm + v;
+    }
+    DualQuat<T> coeff = DualQuat<T>(w, 0, 0, 0, w_, 0, 0, 0) / qNorm;
+    DualQuat<T> k = DualQuat<T>{std::acos(coeff.w), 0, 0, 0, -coeff.w_ / std::sqrt(1 - coeff.w * coeff.w), 0, 0, 0} / normV;
+    return log_qNorm + v * k;
+}
```

Comment:
Now I think to use Taylor expansion is better.

Reply:
If you still want to use Taylor expansion for log, I think the [Inverse function theorem](https://en.wikipedia.org/wiki/Inverse_function_theorem) can be used here to get a jacobian of log:
```
J_log_quat(y=exp(x)) = J_exp_quat(x)^-1
J_log_quat(y) = J_exp_quat(x=log(y))^-1
```
Here log can be calculated with `assumeUnit` argument.

I have calculated an inverse of exp jacobian which can be useful. Please check it if you want to use it.

```
nv = norm(v)
cisii(x) = (cos(x) - x/sin(x))/x/x
J_inv(v) =
[     cos(nv);   v^T*sinc(nv)
  -v*sinc(nv);   I_3*1/sinc(nv) + v*v^T * cisii(nv) ]
p = log(q)
J_log_quat(q) = 1/exp(p.w)*J_inv(p.xyz)
log(w_real+e*w_dual)  = log(w_real) + e*J_log_quat(w_real)*w_dual
```

Here's scipy code I used to get it:
<details>

```python
import sympy
#forgot w, exp(w), 1/exp(w)

#w, x, y, z = log(w_real)

w, x, y, z = sympy.symbols('w x y z')
symb = False
if symb:
    nv = sympy.Symbol('nv')
    cs = sympy.Symbol('cs')
    sinc, csiii = sympy.symbols('sinc csiii')
else:
    nv = sympy.Symbol('nv')
    cs = sympy.cos(nv)
    sinc = sympy.sin(nv)/nv
    csiii = (sympy.cos(nv)-sinc)/nv/nv

vt = sympy.Matrix([[0, x, y, z], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]])
vvt = vt.T * vt

I_3 = sympy.Matrix([[0, 0, 0, 0], [0, 1, 0, 0], [0, 0, 1, 0], [0, 0, 0, 1]])
cosv = sympy.Matrix([[cs, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]])

wpart = cosv + sinc*vt.T
ijkpart = -sinc*vt + vvt * csiii + sinc * I_3

jexp = wpart+ijkpart

# jexp.det() == (sympy.sin(nv)/nv)**2

# this takes too long but works
jinv = jexp ** -1

nvsym = sympy.sqrt(x*x + y*y + z*z)

jinv_simp = sympy.Matrix([[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]])
for i in range(4):
    for j in range(4):
        jinv_simp[i, j] = jinv[i, j].subs(nv, nvsym).simplify()
        
# constructing

jinv_s = jinv_simp.subs(nvsym, nv)
invsinc = 1/sinc

# here it is, don't forget 1/exp(w)*jinv_c
jinv_c = cosv + vt*sinc - vt.T*sinc + I_3*invsinc + vvt * (sympy.cos(nv) - invsinc) / nv**2

jinv_comp = jinv_c.subs(nv, nvsym)

diff = jinv_simp - jinv_comp

for i in range(4):
    for j in range(4):
        diff[i, j] = diff[i, j].simplify()
        
diff # output is 4x4 zeros
```

</details>